accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,134 @@
+
+For the *language versioning* feature, we have proposed extending the format by adding fragments to the URI references and an entry with the empty string as key. This is a breaking change (the empty key was not previously accepted) and even more breaking for tools which assume the values are paths.
+
+Instead of breaking the existing file users, we will instead introduce a *new* file containing the information that we intended to store in the `.packages` file. This fill will use a (more) structured format which allows extending it with further information in the future. We will still generate the `.packages` file for now, allowing users of the file to continue working until they can migrate to using the new file.
+
+## New File Contents
+
+The file will contain the information that was planned for the `.packages` file:
+
+- A *default package* specified by its package name.
+- A list of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI*.
+  - A *language version* (major/minor version number)
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+We use JSON for the format. JSON has parsers available for most languages (not all Dart tools are written in Dart) and it is standardized and efficiently parsable.  JSON is *not* easy to write by hand (but doable), and it does not allow comments, but this is a machine generated file that will hardly ever be edited by hand.
+
+## New File Format
+
+The new file will be called `.packages.json` and stored in the same directory as the `.packages` file. The format is JSON text describing is a single JSON object.
+
+That object has, at least, the following properties:
+
+- `apiVersion`: A single integer marking the version of this format, currently `2` (counting `.packages` as the first). Tools should refuse to work on files with a larger version number than the ones they are aware of. Incremental changes will not require increasing the version, so new versions are not expected any time soon, but code should still prepare for it.
+
+- `defaultPackage`: A string containing a package name.
+- `packages`: A list of packages, each with the following properties:
+  - `name`: A string containing a package name.
+  - either `path`: A string containing a relative or absolute path using `/` as separator. Every character other than `/` is used verbatim. If the path starts with a `/` or with a DOS/Windows drive letter (a single letter followed by `:/`), then it is considered absolute. If not, the path is relative to the directory of the `.packages.json` file. 
+  - or `uri`: A URI reference. This is resolved relative to the (likely `file:`) URI referencing the `.package.json` file. If the URI reference has no scheme, or if it has a `file:` scheme, it is recommended to use the `path` instead.
+  - `languageVersion`: Optional. A string containing a language version (major.minor).
```

Comment:
You'll need to specify the version more precisely. Alternatively, if you don't want to embed a mini language inside one of the JSON fields, you could do something like:

```json
{
  "language": {
    "version": {
      "major": 1,
      "minor": 2
    }
  }
}
```

That's pretty gross. :( But it does increase the odds of any random tool correctly parsing a version number. Also, it avoids users/tools accidentally writing:

```json
"languageVersion": 1.2
```

Which definitely does *not* mean what they think it means.

Also, I think "dart" is more useful than "language" here. We could conceivably extend this to put stuff related to other languages in here, but "Dart" is pretty precise.

Reply:
I loathe unnecessary verbosity, so let's define the format:

> A string containing a decimal numeral, a `.` and another decimal numeral, where the decimal numerals have no leading spaces unless the numeral is precisely `0`. Any string matched in its entirety by the RegExp `(0|[1-9]\d*)\.(0|[1-9]\d*)`.

That's how we specify language version inside Dart files in `// @dart = x.y` comments, and it's also a valid major/minor version of a SemVer version.

We probably won't accept double entries at all, not because of `1.2`, but because of `1.20`. It must be a string.